alchemiscale/storage/statestore.py

```diff
@@ -1463,56 +1469,260 @@ def get_task_failures(self, task: ScopedKey) -> List[ProtocolDAGResultRef]:
         """
         return self._get_protocoldagresultrefs(q)
 
-    def set_task_waiting(self, task: ScopedKey, clear_claim=True):
-        q = f"""
-        MATCH (t:Task {{_scoped_key: "{task}"}})
-        SET t.status = 'waiting'
+    def set_task_status(
+        self, tasks: List[ScopedKey], status: TaskStatusEnum, raise_error: bool = False
+    ) -> List[Optional[ScopedKey]]:
+        """Set the status of a list of Tasks.
+
+        This is a master method that calls the appropriate method for the
+        status.
+
+        Parameters
+        ----------
+        tasks
+            The list of Tasks to set the status of.
+        status
+            The status to set the Task to.
+        raise_error
+            If `True`, raise a `ValueError` if the status of a given Task cannot be changed.
+
+        Returns
+        -------
+        List[Optional[ScopedKey]]
+            A list of the Task ScopedKeys for which status was changed; `None`
+            is given for any Tasks for which the status could not be changed.
+
         """
+        method = getattr(self, f"set_task_{status.value}")
+        return method(tasks, raise_error=raise_error)
 
-        if clear_claim:
-            q += ", t.claimant = null"
+    def get_task_status(self, tasks: List[ScopedKey]) -> List[TaskStatusEnum]:
+        """Get the status of a list of Tasks.
 
-        q += """
-        RETURN t
+        Parameters
+        ----------
+        tasks
+            The list of Tasks to get the status for.
+
+        Returns
+        -------
+        Dict[ScopedKey,TaskStatusEnum]
+            A dictionary of Tasks and their statuses.
         """
 
+        statuses = []
         with self.transaction() as tx:
-            tx.run(q)
+            for t in tasks:
+                q = f"""
+                MATCH (t:Task {{_scoped_key: '{t}'}})
+                RETURN t
+                """
+                task = tx.run(q).to_subgraph()
+                if task is None:
+                    statuses.append(None)
+                else:
+                    status = task.get("status")
+                    statuses.append(TaskStatusEnum(status))
 
-    def set_task_running(self, task: ScopedKey, computekey: ComputeKey):
-        ...
+        return statuses
+
+    def _set_task_status(self, tasks, q_func, err_msg_func, raise_error):
+        tasks_set = []
+        with self.transaction() as tx:
+            for t in tasks:
+                res = tx.run(q_func(t))
+                for record in res:
```

Comment:
If its not possible to have more than one record by design, perhaps leave a comment or use some other syntax  other than a `for` loop that makes it clearer?

Reply:
Ah, good point; I added an exception that will get raised if more than one record shows up. It should not be possible.

Reply:
Went for a `break` and a comment instead. Under certain conditions we will get more than one record back; I believe it has to do with whether any additional `OPTIONAL MATCH` or `MATCH` statements end up triggering. Because of the constraint on these objects, however, we will never have actual duplicate nodes in the DB.